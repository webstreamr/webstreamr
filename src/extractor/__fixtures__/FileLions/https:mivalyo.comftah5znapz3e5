<!DOCTYPE html>
<html lang="en">

<head>
  <meta charset="utf-8">
  <title>
    
      Watch Versailles S01E01 German AC3 1080p BDRip x265 FuN
        
  </title>
    <META NAME="description" CONTENT="">
    <META NAME="keywords" CONTENT="">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
  <meta name="theme-color" content="#FFF">
  
  <link rel="apple-touch-icon" href="/static/images/favicon/apple-touch-icon.png" sizes="180x180">
  <link rel="icon" href="/static/images/favicon/favicon-32x32.png" sizes="32x32">
  <link rel="icon" href="/static/images/favicon/favicon-16x16.png" sizes="16x16">
  <link rel="manifest" href="/static/images/favicon/manifest.json">
  <link rel="icon" href="/static/images/favicon/favicon.ico">
  <link rel="preconnect" href="https://fonts.googleapis.com">
  <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin="crossorigin">
  <link href="https://fonts.googleapis.com/css2?family=Poppins:wght@400;500;700&display=swap" rel="stylesheet">
  <link href="/static/css/style.css?" rel="stylesheet">

  <script src="/static/js/jquery-3.2.1.min.js?"></script>
  <script src="/static/js/bootstrap.bundle.min.js?"></script>
  <script src="/static/js/app.js?"></script>
  <script src="/static/js/xupload.js?"></script>


    <!-- Google tag (gtag.js) -->
<script async src="https://www.googletagmanager.com/gtag/js?id=G-48ZJD1VPGZ"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-48ZJD1VPGZ');
</script>
<!-- Yandex.Metrika counter -->
<script type="text/javascript" >
  (function(m,e,t,r,i,k,a){m[i]=m[i]||function(){(m[i].a=m[i].a||[]).push(arguments)};
  m[i].l=1*new Date();
  for (var j = 0; j < document.scripts.length; j++) {if (document.scripts[j].src === r) { return; }}
  k=e.createElement(t),a=e.getElementsByTagName(t)[0],k.async=1,k.src=r,a.parentNode.insertBefore(k,a)})
  (window, document, "script", "https://mc.yandex.ru/metrika/tag.js", "ym");

  ym(93381808, "init", {
       clickmap:true,
       trackLinks:true,
       accurateTrackBounce:true,
       webvisor:true
  });
</script>
<noscript><div><img src="https://mc.yandex.ru/watch/93381808" style="position:absolute; left:-9999px;" alt="" /></div></noscript>
<!-- /Yandex.Metrika counter -->
</head>

<body>

  <header class="header">
    <div class="container">
      <div class="row align-items-center flex-nowrap">
        <div class="col-auto flex-grow-1 flex-shrink-1">
          <div class="row align-items-center">

            <div class="col-auto" style="margin: 0 auto;">
              <a href="/" class="logo">
                
                    <img src="/static/images/logo.svg" alt="">
                
              </a>
            </div>

          </div>
        </div>


      </div>
    </div>
  </header>


  <main>
  <section class="section">
    <div class="container">
    
      <div class="alert alert-danger">
        This video can be watched as embed only
      </div>
    
    
  </div>
</section>
</main>
<footer class="footer">
  <div class="container">
    <div class="row">
      <div class="col-lg-4 col-xs-12">
        <div class="mb-3 text-center">
          <a href="#" class="logo">
            
                <img src="/static/images/logo.svg" alt="">
            
          </a>
        </div>

      </div>

      <div class="col-lg-8 col-xs-12" style="position: relative">
        <div class="mb-4 text-center">
          <p class="text-muted ">
            
          </p>
        </div>
      </div>
    </div>
  </div>
</footer>
<style>
  @media (min-width:481px) {
    .text-muted {}
  }

  @media (min-width:961px) {
    .text-muted {
      position: absolute;
      right: 0;
      top: 15px
    }
  }
</style>
</body>

</html>

