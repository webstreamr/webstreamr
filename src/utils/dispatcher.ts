import { socksDispatcher } from 'fetch-socks';
import { minimatch } from 'minimatch';
import { Dispatcher, ProxyAgent } from 'undici';
import { Context } from '../types';

export const getProxyForUrl = (ctx: Context, url: URL): URL | undefined => {
  const proxyConfig = ctx.config['proxyConfig'] || process.env['PROXY_CONFIG'];

  if (proxyConfig) {
    for (const rule of proxyConfig.split(',')) {
      const [hostPattern, proxy] = rule.split(/:(.+)/);
      if (!hostPattern || !proxy) {
        throw new Error(`Proxy rule "${rule}" is invalid.`);
      }

      if (hostPattern === '*' || minimatch(url.host, hostPattern)) {
        return proxy === 'false' ? undefined : new URL(proxy);
      }
    }
  } else if (process.env['ALL_PROXY']) {
    return new URL(process.env['ALL_PROXY']);
  }

  return undefined;
};

export const createProxyAgent = (proxyUrl: URL): Dispatcher => {
  if (proxyUrl.protocol === 'socks5:') {
    return socksDispatcher({ type: 5, host: proxyUrl.hostname, port: parseInt(proxyUrl.port) });
  }

  return new ProxyAgent({ uri: proxyUrl.href });
};
