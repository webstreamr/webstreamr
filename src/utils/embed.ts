import { unpack } from 'unpacker';

export class CouldNotFindStreamLinkInEmbedError extends Error {}

export const extractUrlFromPacked = (html: string, linkRegExps: RegExp[]): URL => {
  const evalMatch = html.match(/eval\(function\(p,a,c,k,e,d\).*\)\)/);
  if (!evalMatch) {
    throw new Error(`No p.a.c.k.e.d string found`);
  }

  const unpacked = unpack(evalMatch[0]);

  for (const linkRegexp of linkRegExps) {
    const linkMatch = unpacked.match(linkRegexp);
    if (linkMatch && linkMatch[1]) {
      return new URL('https://' + linkMatch[1].replace(/^(https:)?\/\//, ''));
    }
  }

  throw new CouldNotFindStreamLinkInEmbedError(`Could not find a stream link in embed`);
};
