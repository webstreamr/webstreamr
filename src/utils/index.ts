export * from './embed';
export * from './fetch';
export * from './imdb';
export * from './language';
export * from './log';
export * from './manifest';
export * from './promise';
