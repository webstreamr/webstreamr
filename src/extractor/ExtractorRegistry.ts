// eslint-disable-next-line import/no-named-as-default
import KeyvSqlite from '@keyv/sqlite';
import { Cacheable, CacheableMemory, Keyv } from 'cacheable';
import winston from 'winston';
import { Context, Meta, UrlResult } from '../types';
import { getCacheDir, isExtractorDisabled } from '../utils';
import { Extractor } from './Extractor';

export class ExtractorRegistry {
  private readonly logger: winston.Logger;
  private readonly extractors: Extractor[];
  private readonly urlResultCache: Cacheable;

  public constructor(logger: winston.Logger, extractors: Extractor[]) {
    this.logger = logger;
    this.extractors = extractors;

    this.urlResultCache = new Cacheable({
      primary: new Keyv({ store: new CacheableMemory({ lruSize: 1024 }) }),
      secondary: new Keyv(new KeyvSqlite(`sqlite://${getCacheDir()}/webstreamr-extractor-cache.sqlite`)),
      stats: true,
    });
  }

  public stats() {
    return {
      urlResultCache: this.urlResultCache.stats,
    };
  };

  public async handle(ctx: Context, url: URL, meta?: Meta): Promise<UrlResult[]> {
    const extractor = this.extractors.find(extractor => !isExtractorDisabled(ctx.config, extractor) && extractor.supports(ctx, url));
    if (!extractor) {
      return [];
    }

    const normalizedUrl = extractor.normalize(url);
    const cacheKey = extractor.viaMediaFlowProxy
      ? `${extractor.id}_${normalizedUrl}_${ctx.config.mediaFlowProxyUrl}`
      : `${extractor.id}_${normalizedUrl}`;

    const storedDataRaw = await this.urlResultCache.get<UrlResult[]>(cacheKey, { raw: true });
    const expires = storedDataRaw?.expires;
    if (storedDataRaw && expires) {
      // Ignore the cache randomly after at least 2/3 of the TTL passed to start refreshing results slowly
      const refreshTimestamp = this.randomInteger(expires - extractor.ttl * (2 / 3), expires);
      const now = Date.now();

      /* istanbul ignore if */
      if (refreshTimestamp > now) {
        return (storedDataRaw.value as UrlResult[]).map(urlResult => ({ ...urlResult, ttl: expires - now, url: new URL(urlResult.url) }));
      }
    }

    this.logger.info(`Extract stream URL using ${extractor.id} extractor from ${url}`, ctx);

    const urlResults = await extractor.extract(ctx, normalizedUrl, { ...meta, countryCodes: meta?.countryCodes ?? [] });

    if (!urlResults.some(urlResult => urlResult.error) && extractor.ttl) {
      await this.urlResultCache.set<UrlResult[]>(cacheKey, urlResults, extractor.ttl);
    }

    return urlResults;
  };

  private randomInteger(min: number, max: number) {
    return Math.floor(Math.random() * (max - min + 1)) + min;
  }
}
