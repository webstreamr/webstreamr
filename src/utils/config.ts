import { Extractor } from '../extractor';
import { Config } from '../types';

export const getDefaultConfig = (): Config => {
  return { multi: 'on', en: 'on' };
};

export const showErrors = (config: Config): boolean => 'showErrors' in config;

export const showExternalUrls = (config: Config): boolean => 'includeExternalUrls' in config;

export const hasMultiEnabled = (config: Config): boolean => 'multi' in config;

export const disableExtractorConfigKey = (extractor: Extractor): string => `disableExtractor_${extractor.id}`;

export const isExtractorDisabled = (config: Config, extractor: Extractor): boolean => disableExtractorConfigKey(extractor) in config;
