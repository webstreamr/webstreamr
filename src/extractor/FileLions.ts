import bytes from 'bytes';
import * as cheerio from 'cheerio';
import { Context, Format, Meta, UrlResult } from '../types';
import { buildMediaFlowProxyExtractorStreamUrl, guessHeightFromPlaylist, supportsMediaFlowProxy } from '../utils';
import { Extractor } from './Extractor';

/** @see https://github.com/Gujal00/ResolveURL/commits/master/script.module.resolveurl/lib/resolveurl/plugins/filelions.py */
export class FileLions extends Extractor {
  public readonly id = 'filelions';

  public readonly label = 'FileLions (via MediaFlow Proxy)';

  public override readonly ttl = 0;

  public override viaMediaFlowProxy = true;

  public supports(ctx: Context, url: URL): boolean {
    const supportedDomain = null !== url.host.match(/.*lions?/)
      || [
        '6sfkrspw4u.sbs',
        'ajmidyadfihayh.sbs',
        'alhayabambi.sbs',
        'anime7u.com',
        'azipcdn.com',
        'coolciima.online',
        'dhtpre.com',
        'e4xb5c2xnz.sbs',
        'egsyxutd.sbs',
        'fdewsdc.sbs',
        'gsfomqu.sbs',
        'javplaya.com',
        'katomen.online',
        'kinoger.be',
        'lumiawatch.top',
        'mivalyo.com',
        'moflix-stream.click',
        'motvy55.store',
        'movearnpre.com',
        'peytonepre.com',
        'ryderjet.com',
        'smoothpre.com',
        'taylorplayer.com',
        'techradar.ink',
        'videoland.sbs',
        'vidhide.com',
        'vidhide.fun',
        'vidhidefast.com',
        'vidhidehub.com',
        'vidhideplus.com',
        'vidhidepre.com',
        'vidhidepro.com',
        'vidhidevip.com',
      ].includes(url.host);

    return supportedDomain && supportsMediaFlowProxy(ctx);
  }

  public override normalize(url: URL): URL {
    return new URL(url.href.replace('/f/', '/v/').replace('/download/', '/v/').replace('/file/', '/v/'));
  }

  protected async extractInternal(ctx: Context, url: URL, meta: Meta): Promise<UrlResult[]> {
    const headers = { ...(meta.referer && { Referer: meta.referer }) };

    const playlistUrl = await buildMediaFlowProxyExtractorStreamUrl(ctx, this.fetcher, 'FileLions', url, headers);

    const fileUrl = new URL(url.href.replace('/v/', '/f/'));
    const html = await this.fetcher.text(ctx, fileUrl, { headers });

    const sizeMatch = html.match(/([\d.]+ ?[GM]B)/);

    const $ = cheerio.load(html);
    const title = $('title').text().trim().replace(/^Watch /, '').trim();

    return [
      {
        url: playlistUrl,
        format: Format.hls,
        label: this.label,
        sourceId: `${this.id}_${meta.countryCodes?.join('_')}`,
        ttl: this.ttl,
        meta: {
          ...meta,
          height: await guessHeightFromPlaylist(ctx, this.fetcher, playlistUrl),
          ...(sizeMatch && {
            bytes: bytes.parse(sizeMatch[1] as string) as number,
          }),
          ...(title && { title }),
        },
      },
    ];
  };
}
