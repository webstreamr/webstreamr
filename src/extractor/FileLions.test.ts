import winston from 'winston';
import { createTestContext } from '../test';
import { FetcherMock } from '../utils';
import { ExtractorRegistry } from './ExtractorRegistry';
import { FileLions } from './FileLions';

const logger = winston.createLogger({ transports: [new winston.transports.Console({ level: 'nope' })] });
const extractorRegistry = new ExtractorRegistry(logger, [new FileLions(new FetcherMock(`${__dirname}/__fixtures__/FileLions`))]);

const ctx = createTestContext({ mediaFlowProxyUrl: 'https://mediaflow.test.org', mediaFlowProxyPassword: 'test' });

describe('FileLions', () => {
  test('filelions f', async () => {
    expect(await extractorRegistry.handle(ctx, new URL('https://filelions.to/f/tyn45apubte2'))).toMatchSnapshot();
  });

  test('mivalyo v referer lock', async () => {
    expect(await extractorRegistry.handle(ctx, new URL('https://mivalyo.com/v/tah5znapz3e5'), { referer: 'https://kinoger.com' })).toMatchSnapshot();
  });
});
