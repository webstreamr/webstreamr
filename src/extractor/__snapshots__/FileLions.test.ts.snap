// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`FileLions filelions f 1`] = `
[
  {
    "format": "hls",
    "label": "FileLions (via MediaFlow Proxy)",
    "meta": {
      "bytes": 481191526,
      "countryCodes": [],
      "height": 720,
      "title": "dexter resurrection s01e01 1080p web h264 successfulcrab",
    },
    "sourceId": "filelions_",
    "ttl": 0,
    "url": "https://mediaflow.test.org/proxy/hls/manifest.m3u8?api_password=test&h_user-agent=Mozilla%2F5.0+%28Windows+NT+10.0%3B+Win64%3B+x64%29+AppleWebKit%2F537.36+%28KHTML%2C+like+Gecko%29+Chrome%2F136.0.0.0+Safari%2F537.36&h_referer=https%3A%2F%2Ffilelions.to%2Fv%2Ftyn45apubte2&d=https%3A%2F%2FRNzT2t4XVKU08.acek-cdn.com%2Fhls2%2F01%2F05594%2Ftyn45apubte2_n%2Fmaster.m3u8%3Ft%3D8y6hK40boUynXnkyW5s63kFaWpt6_A5UcG5LBUX3j_4%26s%3D1756131487%26e%3D129600%26f%3D27974684%26srv%3DF6vE2P5Wjncn%26i%3D0.4%26sp%3D500%26p1%3DF6vE2P5Wjncn%26p2%3DF6vE2P5Wjncn%26asn%3D31898",
  },
]
`;

exports[`FileLions mivalyo v referer lock 1`] = `
[
  {
    "format": "hls",
    "label": "FileLions (via MediaFlow Proxy)",
    "meta": {
      "countryCodes": [],
      "height": 1080,
      "referer": "https://kinoger.com",
      "title": "Versailles S01E01 German AC3 1080p BDRip x265 FuN",
    },
    "sourceId": "filelions_",
    "ttl": 0,
    "url": "https://mediaflow.test.org/proxy/hls/manifest.m3u8?api_password=test&h_user-agent=Mozilla%2F5.0+%28Windows+NT+10.0%3B+Win64%3B+x64%29+AppleWebKit%2F537.36+%28KHTML%2C+like+Gecko%29+Chrome%2F136.0.0.0+Safari%2F537.36&h_referer=https%3A%2F%2Fkinoger.com&d=https%3A%2F%2F6dkDxt1udLQhbho.ovaltinecdn.com%2Fhls2%2F01%2F01552%2Ftah5znapz3e5_n%2Fmaster.m3u8%3Ft%3DMnSHt6lZx3bdKvTdoU1JsTLWp56dlRDLwze5Prnefew%26s%3D1756131487%26e%3D129600%26f%3D7760835%26srv%3DGOweneve6B0I%26i%3D0.4%26sp%3D500%26p1%3DGOweneve6B0I%26p2%3DGOweneve6B0I%26asn%3D31898",
  },
]
`;
